src/zarr/core/metadata/dtype.py

```diff
@@ -0,0 +1,806 @@
+
+    def to_dict(self) -> dict[str, JSON]:
+        return {"name": self.name}
+
+    def cast_value(self: Self, value: object) -> TScalar:
+        return cast(TScalar, self.unwrap().type(value))
+
+    @abstractmethod
+    def default_value(self) -> TScalar: ...
+
+    @classmethod
+    def check_dtype(cls: type[Self], dtype: TDType) -> TypeGuard[TDType]:
+        """
+        Check that a dtype matches the dtype_cls class attribute
+        """
+        return type(dtype) is cls.dtype_cls
+
+    @classmethod
+    def wrap(cls: type[Self], dtype: TDType) -> Self:
+        if cls.check_dtype(dtype):
+            return cls._wrap_unsafe(dtype)
+        raise TypeError(f"Invalid dtype: {dtype}. Expected an instance of {cls.dtype_cls}.")
+
+    @classmethod
+    @abstractmethod
+    def _wrap_unsafe(cls: type[Self], dtype: TDType) -> Self:
+        raise NotImplementedError
+
+    def unwrap(self: Self) -> TDType:
+        endian_str = endianness_to_numpy_str(self.endianness)
+        return self.dtype_cls().newbyteorder(endian_str)
+
+    def with_endianness(self: Self, endianness: Endianness) -> Self:
+        return replace(self, endianness=endianness)
+
+    @abstractmethod
+    def to_json_value(self, data: np.generic, *, zarr_format: ZarrFormat) -> JSON:
+        """
+        Convert a single value to JSON-serializable format. Depends on the zarr format.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def from_json_value(self: Self, data: JSON, *, zarr_format: ZarrFormat) -> TScalar:
+        """
+        Read a JSON-serializable value as a numpy scalar
+        """
+        raise NotImplementedError
+
+
+@dataclass(frozen=True, kw_only=True)
+class Bool(DTypeWrapper[np.dtypes.BoolDType, np.bool_]):
+    name = "bool"
+
+    def default_value(self) -> np.bool_:
+        return np.False_
+
+    @classmethod
+    def _wrap_unsafe(cls, dtype: np.dtypes.BoolDType) -> Self:
+        return cls()
+
+    def to_json_value(self, data: np.generic, zarr_format: ZarrFormat) -> bool:
+        return bool(data)
+
+    def from_json_value(self, data: JSON, *, zarr_format: ZarrFormat) -> np.bool_:
+        if check_json_bool(data):
+            return self.unwrap().type(data)
+        raise TypeError(f"Invalid type: {data}. Expected a boolean.")
+
+
+class IntWrapperBase(DTypeWrapper[TDType, TScalar]):
+    def default_value(self) -> TScalar:
+        return self.unwrap().type(0)
+
+    @classmethod
+    def _wrap_unsafe(cls, dtype: TDType) -> Self:
+        return cls()
+
+    def to_json_value(self, data: np.generic, zarr_format: ZarrFormat) -> int:
+        return int(data)
+
+    def from_json_value(self, data: JSON, *, zarr_format: ZarrFormat) -> TScalar:
+        if check_json_int(data):
+            return self.unwrap().type(data)
+        raise TypeError(f"Invalid type: {data}. Expected an integer.")
+
+
+@dataclass(frozen=True, kw_only=True)
+class Int8(IntWrapperBase[np.dtypes.Int8DType, np.int8]):
+    name = "int8"
```

Comment:
Perhaps all the concrete definitions of dtypes that zarr natively supports should live in a separate file from the definitions of the framework that allows it to support custom dtypes?

That might help prevent the pseudo-auto-generated code from crowding out the much more important definitions.

Reply:
that's a good idea! I will make that change